File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lyricsscraper/dialog/CacheDialogFragment.java
package com.wa2c.android.medoly.plugin.action.lyricsscraper.dialog;

import android.app.AlertDialog;
import android.app.Dialog;
import android.os.AsyncTask;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.TextView;


import com.wa2c.android.medoly.plugin.action.lyricsscraper.R;
import com.wa2c.android.medoly.plugin.action.lyricsscraper.db.SearchCache;
import com.wa2c.android.medoly.plugin.action.lyricsscraper.db.SearchCacheHelper;
import com.wa2c.android.medoly.plugin.action.lyricsscraper.search.ResultItem;

import java.util.Arrays;


/**
 * Confirm dialog.
 */
public class CacheDialogFragment extends AbstractDialogFragment {

    /** Dialog result on delete lyrics. */
    public final static int DIALOG_RESULT_DELETE_LYRICS = -10;
    /** Dialog result on delete cache. */
    public final static int DIALOG_RESULT_DELETE_CACHE = -20;

    /** Cache key. */
    private final static String ARG_CACHE = "ARG_CACHE";

    /**
     * Create dialog instance.
     * @param cache Search cache.
     * @return Dialog instance.
     */
    static public CacheDialogFragment newInstance(SearchCache cache) {
        CacheDialogFragment fragment = new CacheDialogFragment();
        Bundle args = new Bundle();
        args.putSerializable(ARG_CACHE, cache);
        fragment.setArguments(args);

        return fragment;
    }



    /**
     * onCreateDialog
     */
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        super.onCreateDialog(savedInstanceState);

        // data
        final SearchCache cache = (SearchCache)getArguments().getSerializable(ARG_CACHE);
        ResultItem result = cache.makeResultItem();

        // view
        final View content = View.inflate(getActivity(), R.layout.dialog_cache, null);
        TextView textView = (TextView)content.findViewById(R.id.dialogCacheLyricsTextView);
        textView.setText((cache.has_lyrics != null && cache.has_lyrics) ? result.getLyrics() : getString(R.string.message_dialog_cache_none));

//        // delete lyrics button
//        content.findViewById(R.id.dialogCacheDeleteLyricsButton).setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                (new AsyncTask<Void, Void, Void>() {
//                    @Override
//                    protected Void doInBackground(Void... params) {
//                        SearchCacheHelper searchCacheHelper = new SearchCacheHelper(getActivity());
//                        searchCacheHelper.insertOrUpdate(cache.title, cache.artist, null);
//                        return null;
//                    }
//                    @Override
//                    protected void onPostExecute(Void result) {
//                        onClickButton(getDialog(), DIALOG_RESULT_DELETE_LYRICS);
//                    }
//                }).execute();
//             }
//        });
//        // delete cache button
//        content.findViewById(R.id.dialogCacheDeleteCacheButton).setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                (new AsyncTask<Void, Void, Void>() {
//                    @Override
//                    protected Void doInBackground(Void... params) {
//                        SearchCacheHelper searchCacheHelper = new SearchCacheHelper(getActivity());
//                        searchCacheHelper.delete(Arrays.asList(cache));
//                        return null;
//                    }
//                    @Override
//                    protected void onPostExecute(Void result) {
//                        onClickButton(getDialog(), DIALOG_RESULT_DELETE_CACHE);
//                    }
//                }).execute();
//            }
//        });

        // build dialog
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        builder.setTitle(R.string.title_activity_cache);
        builder.setView(content);
        builder.setNeutralButton(R.string.label_close, null);
        builder.setNegativeButton(R.string.label_dialog_cache_research, clickListener);
        if (result != null && !TextUtils.isEmpty(result.getLyrics())) {
            builder.setPositiveButton(R.string.menu_search_save_file, clickListener);
        }
        return  builder.create();
    }

}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lyricsscraper/dialog/CacheDialogFragment.kt
package com.wa2c.android.medoly.plugin.action.lyricsscraper.dialog

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import androidx.appcompat.app.AlertDialog
import androidx.databinding.DataBindingUtil
import com.wa2c.android.medoly.plugin.action.lyricsscraper.R
import com.wa2c.android.medoly.plugin.action.lyricsscraper.databinding.DialogCacheBinding
import com.wa2c.android.medoly.plugin.action.lyricsscraper.db.DbHelper
import com.wa2c.android.medoly.plugin.action.lyricsscraper.db.SearchCache
import com.wa2c.android.medoly.plugin.action.lyricsscraper.util.AppUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch

/**
 * Cache dialog.
 */
class CacheDialogFragment : AbstractDialogFragment() {

    private lateinit var binding: DialogCacheBinding

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        super.onCreateDialog(savedInstanceState)
        binding = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.dialog_cache, null, false)

        // data
        val cache = arguments?.getSerializable(ARG_CACHE) as SearchCache
        val result = cache.makeResultItem()

        // view
        binding.dialogCacheLyricsTextView.text = if (cache.has_lyrics != null && cache.has_lyrics!!) result!!.lyrics else getString(R.string.message_dialog_cache_none)
        binding.dialogCacheDeleteLyricsButton.isEnabled = (cache.has_lyrics == true)

        // deleteCache lyrics button
        binding.dialogCacheDeleteLyricsButton.setOnClickListener {
            deleteLyrics(cache)
        }

        // deleteCache cache button
        binding.dialogCacheDeleteCacheButton.setOnClickListener {
            deleteCache(cache)
        }

        // build dialog
        val builder = AlertDialog.Builder(context)
        builder.setTitle(R.string.title_activity_cache)
        builder.setView(binding.root)
        builder.setNeutralButton(R.string.label_close, null)
        builder.setNegativeButton(R.string.label_dialog_cache_research, null)
        if (result != null && !result.lyrics.isNullOrEmpty()) {
            builder.setPositiveButton(R.string.menu_search_save_file, null)
        }
        return builder.create()
    }

    /**
     * Delete cache lyrics.
     */
    private fun deleteLyrics(cache: SearchCache) {
        val searchCacheHelper = DbHelper(context)
        GlobalScope.launch(Dispatchers.Main) {
            val deleteResult = async(Dispatchers.Default) {
                try {
                    searchCacheHelper.insertOrUpdateCache(cache.title, cache.artist, null)
                } catch (e: Exception) {
                    return@async null
                }
            }
            val r = deleteResult.await()
            if (r != true)
                AppUtils.showToast(context, R.string.message_dialog_cache_delete_error)
            clickListener?.invoke(dialog, DIALOG_RESULT_DELETE_LYRICS, null)
            dialog.dismiss()
        }
    }

    /**
     * Delete cache.
     */
    private fun deleteCache(cache: SearchCache) {
        val searchCacheHelper = DbHelper(context)
        GlobalScope.launch(Dispatchers.Main) {
            val deleteResult = async(Dispatchers.Default) {
                try {
                    searchCacheHelper.deleteCache(listOf(cache))
                } catch (e: Exception) {
                    return@async null
                }
            }
            val r = deleteResult.await()
            if (r != true)
                AppUtils.showToast(context, R.string.message_dialog_cache_delete_error)
            clickListener?.invoke(dialog, DIALOG_RESULT_DELETE_CACHE, null)
            dialog.dismiss()
        }
    }


    companion object {

        /** Dialog result on deleteCache lyrics.  */
        const val DIALOG_RESULT_DELETE_LYRICS = -10
        /** Dialog result on deleteCache cache.  */
        const val DIALOG_RESULT_DELETE_CACHE = -20

        /** Cache key.  */
        private const val ARG_CACHE = "ARG_CACHE"

        /**
         * Create dialog instance.
         * @param cache Search cache.
         * @return Dialog instance.
         */
        fun newInstance(cache: SearchCache): CacheDialogFragment {
            val fragment = CacheDialogFragment()
            val args = Bundle()
            args.putSerializable(ARG_CACHE, cache)
            fragment.arguments = args

            return fragment
        }
    }
}
